### PR TITLE
[cmake] Packaging for OSX

### PR DESCRIPTION
**Pull request by fetzerch**
This introduces a `osx_makedmg` target that packages a dmg using the already existing shell scripts (copyframeworks, copyroot, dmgmaker, ...). This is only an intermediate step because CPack would be the proper way to do it. That'll unfortunately take a bit more time, but can easily be done later. Because of this, I didn't want to modify the existing scripts (such as hardcoded relative paths to the source build) but tried to workaround issues in the install.cmake.

Jenkins bits are in place as well so that a [first dmg](http://mirrors.kodi.tv/test-builds/osx/x86_64/kodi-20160430-48921b4-cmake_osx_packaging-x86_64.dmg) can be downloaded ([job](http://jenkins.kodi.tv/job/OSX-64/8138/)). Kodi starts up, binary addons are usable and except empty folders, `.gitignore`s, `Makefiles`, `addon.xml.in` which we don't package with CMake, the image contains the same content.

@Memphiz: This is probably interesting for you

### CODE REVIEW

File: tools/darwin/packaging/osx/mkdmg-osx.sh.in [hunk 1]
```diff
@@ -34,7 +34,7 @@ fi
 ARCHIVE=${PACKAGE}_${VERSION}-${REVISION}_macosx-intel-${ARCHITECTURE}
 
 echo Creating $PACKAGE package version $VERSION revision $REVISION
-${SUDO} rm -rf $DIRNAME/$ARCHIVE
+rm -rf $DIRNAME/$ARCHIVE
```
Comment (fetzerch) [inline tools/darwin/packaging/osx/mkdmg-osx.sh.in]: @Memphiz: Surprisingly this failed on jenkins with `sudo: no tty present and no askpass program specified`. Do you have any idea what's causing this and how to fix it? Removing sudo causes issues when rebuilding the image locally. It seems to work on the non-cmake build though.

**Comment by Memphiz** [comment]
Nice! - unsure about the sudo thing - no clue how sudo even works when nobody provides the password - question is - who sets that SUDO var and what is it set to?

**Comment by wsnipex** [comment]
sudo doesn't ask for a password if the sudoers files defines the entry as password-less

**Comment by Memphiz** [comment]
I am 100% that our jenkins slave user is not setup this way ... will try to dig into it ...

**Comment by fetzerch** [comment]
Thanks!, we could probably also just test if the file exists and only then remove it. Then we would not execute sudo on jenkins because the workspace is cleaned anyways. But it would be still interesting to know what causes this.

**Comment by Memphiz** [comment]
@fetzerch i simply echoed the line here:

rm -rf ./org.xbmc.kodi-osx_17.0-0~alpha2_macosx-intel-x86_64

I think SUDO is not set normally ... not sure if it is set in cmake - maybe thats an issue then.

In the osx script we use fakeroot for achieving sudo stuff - see here:

https://github.com/xbmc/xbmc/blob/master/tools/darwin/packaging/ios/mkdeb-ios.sh.in#L43

I think that sudo is a c&p from the ios script tbh and not really needed on osx ...

**Comment by Memphiz** [comment]
or maybe it is needed - in that case we need to move over that fakeroot stuff from the ios script...

**Comment by fetzerch** [comment]
@memphiz: sudo seems to be needed if you build the dmg locally, otherwise a second run cannot remove the already built image. 63ef2ff should work for jenkins then, we'll see: http://jenkins.kodi.tv/job/OSX-64/8163/

**Comment by Memphiz** [comment]
I wonder why your dmf is 68 MB and the nightly 75 MB

**Comment by fetzerch** [comment]
Hm. Good question. The unpacked binary is 20 MB smaller, but I've got no idea what's causing this. Here is a list of symbols that don't exist in the binary compiled with cmake/makefiles: http://paste.ubuntu.com/16195822/. Can it be that they are actually unused but don't get stripped in the xcode build? I read a bit about symbol stripping on osx which seems only to happen when `Deployment postprocessing` is enabled.

**Comment by Memphiz** [comment]
there are a lot of kodi symbols in it - i doubt that we have so much code in our code base that is unused - are you sure that list really shows the missing symbols in the cmake build?

**Comment by Memphiz** [comment]
could you enable that postprocessing stuff in a xcode build and compare if it also reduces the binary that much?

**Comment by fetzerch** [comment]
@Memphiz: I played a bit with both settings: deploy postprocessing (enables stripping global/debug symbols) and dead code removal. With both I was not really able to get a symbol list similar to the one that the cmake build generates. However both drastically reduce size down to ~48 MB (unpacked). I also compared a cmake/xcode build with the cmake/makefile build: Here we have similar sizes (65MB unpacked).

The list above is really the diff. I double checked that (with `nm -a | cut -f3 -d" " | sort -u` and then using `comm`). It's round about 1/6 of the symbols. However I think that this can also be caused by a different way of linking. In the Xcode project for example all platinum source files are directly added to the executable. While with CMake we build lots of small static libs. I'm not sure if it's really worth to debug this deeper at the moment. Shouldn't we rather try out the built program and see what is broken?

New changes:
- The libdvd libs were missing
- SUDO in makedmg is not needed, but it was deleting a nonexistent file and hence a followup build wouldn't succeed (also the unmounting didn't work)
- Simplified packaging. You can now simply run `make dmg` or if you just want the bundle: `make bundle` (dependencies are resolved now, and it's not necessary to call `make all` before)

Triggered another [cmake build](http://jenkins.kodi.tv/job/OSX-64/8181/) and a an additional [xcode build](http://jenkins.kodi.tv/job/OSX-64/8182/) to see if everything is still working.

**Comment by Memphiz** [comment]
thx for digging - really appreciated! - good to know we can slim the binary down that much via xcode tweaking!

File: project/cmake/scripts/darwin/install.cmake [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+configure_file(${CORE_SOURCE_DIR}/tools/darwin/packaging/osx/mkdmg-osx.sh.in
+               ${CMAKE_BINARY_DIR}/tools/darwin/packaging/osx/mkdmg-osx.sh @ONLY)
+
+add_custom_target(dmg
+    COMMAND ${CMAKE_COMMAND} -E copy_directory ${CORE_SOURCE_DIR}/tools/darwin/packaging/osx/
+                                               ${CMAKE_BINARY_DIR}/tools/darwin/packaging/osx/
+    COMMAND ${CMAKE_COMMAND} -E copy_directory ${CORE_SOURCE_DIR}/tools/darwin/packaging/media/osx/
+                                               ${CMAKE_BINARY_DIR}/tools/darwin/packaging/media/osx/
+    COMMAND "SUDO=/usr/bin/sudo" sh ./mkdmg-osx.sh ${CORE_BUILD_CONFIG}
```
Comment (Memphiz) [inline project/cmake/scripts/darwin/install.cmake]: and there was the sudo coming from - good to know that one too :)

Comment (Memphiz) [inline project/cmake/scripts/darwin/install.cmake]: i think this should go now as there is no need to have root power for packaging right?

Comment (fetzerch) [inline project/cmake/scripts/darwin/install.cmake]: Yes, sorry that was a leftover from testing

**Comment by fetzerch** [comment]
@Memphiz: removed sudo. good to go in then?

**Comment by Memphiz** [comment]
Yep - great thx :)

**Pull request merged**